Update DeepSpeed docs (#6528)
PyTorchLightning/pytorch-lightning: docs/source/advanced/multi_gpu.rst

.. code-block:: bash

    pip install deepspeed

If you run into an issue with the install or later in training, ensure that the CUDA version of the pytorch you've installed matches your locally installed CUDA (you can see which one has been recognized by running ``nvcc --version``).

.. note::
    Currently ``resume_from_checkpoint`` and manual optimization are not supported.

    DeepSpeed currently only supports single optimizer, single scheduler within the training loop.

> DeepSpeed ZeRO Stage 2

Reviewer: do we have somewhere also stage 1?

Author: I mention in the first line this includes stage 1! pretty much all the time you'll use stage 2 which includes stage 1 :)

""""""""""""""""""""""

> By default, we enable DeepSpeed ZeRO Stage 2, which partitions your optimizer states (Stage 1) and your gradients (Stage 2) across your GPUs to reduce memory. In most cases, this is more efficient or at parity with DDP, primarily due to the optimized custom communications written by the DeepSpeed team.

Reviewer 2: add a reference link?

> By default we enable DeepSpeed ZeRO Stage 2, which partitions your optimizer states (Stage 1) and your gradients (Stage 2) across your GPUs to reduce memory. In most cases this is more efficient or at parity with DDP, primarily due to the optimized custom communications written by the DeepSpeed team.
> As a result, benefits can also be seen on a single GPU. Do note that the default bucket sizes allocate around ``3.6GB`` of VRAM to use during distributed communications, which can be tweaked when instantiating the plugin described a few sections below.

Reviewer 3: ```suggestion
By default, we enable DeepSpeed ZeRO Stage 2, which partitions your optimizer states (Stage 1) and your gradients (Stage 2) across your GPUs to reduce memory. In most cases, this is more efficient or at parity with DDP, primarily due to the optimized custom communications written by the DeepSpeed team.
As a result, benefits can also be seen on a single GPU. Do note that the default bucket sizes allocate around ``3.6GB`` of VRAM to use during distributed communications, which can be tweaked when instantiating the plugin described in a few sections below.
```

    The plugin sets a reasonable default of ``2e8``, which should work for most low VRAM GPUs (less than ``7GB``), allocating roughly ``3.6GB`` of VRAM as buffer. Higher VRAM GPUs should aim for values around ``5e8``.

For even more speed benefit, DeepSpeed offers an optimized CPU version of ADAM called `DeepSpeedCPUAdam <https://deepspeed.readthedocs.io/en/latest/optimizers.html#adam-cpu>`_ to run the offloaded computation, which is faster than the standard PyTorch implementation.

.. code-block:: python

    import pytorch_lightning
    from pytorch_lightning import Trainer
    from pytorch_lightning.plugins import DeepSpeedPlugin
    from deepspeed.ops.adam import DeepSpeedCPUAdam

    class MyModel(pl.LightningModule):
        ...
        def configure_optimizers(self):
            # DeepSpeedCPUAdam provides 5x to 7x speedup over torch.optim.adam(w)
            return DeepSpeedCPUAdam(self.parameters())

    model = MyModel()
    trainer = Trainer(gpus=4, plugins=DeepSpeedPlugin(cpu_offload=True), precision=16)
    trainer.fit(model)

Custom DeepSpeed Config
"""""""""""""""""""""""

> In some cases you may want to define your own DeepSpeed Config, to access all parameters defined. We've exposed most of the important parameters, however there may be debug parameters to enable. Also, DeepSpeed allows use of custom DeepSpeed optimizers and schedulers defined within a config file which is supported.

Reviewer 3: ```suggestion
In some cases you may want to define your own DeepSpeed Config, to access all parameters defined. We've exposed most of the important parameters, however, there may be debugging parameters to enable. Also, DeepSpeed allows the use of custom DeepSpeed optimizers and schedulers defined within a config file that is supported.
```
